common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,195 @@
+import brave.opentracing.BraveTracer
+import brave.sampler.Sampler
+import io.opentracing.{Span, SpanContext, Tracer}
+import io.opentracing.util.GlobalTracer
+import io.opentracing.propagation.{Format, TextMapExtractAdapter, TextMapInjectAdapter}
+import zipkin2.reporter.{AsyncReporter, Sender}
+import zipkin2.reporter.okhttp3.OkHttpSender
+import pureconfig._
+import whisk.common.{LogMarkerToken, TransactionId}
+import whisk.core.ConfigKeys
+
+import scala.collection.mutable
+import scala.ref.WeakReference
+
+/**
+ * OpenTracing based implementation for tracing
+ */
+class OpenTracer(val tracer: Tracer) extends WhiskTracer {
+
+  /**
+   * Start a Trace for given service.
+   *
+   * @param transactionId transactionId to which this Trace belongs.
+   * @return TracedRequest which provides details about current service being traced.
+   */
+  override def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+
+    //initialize list for this transactionId
+    val spanList: List[WeakReference[Span]] = TracingCacheProvider.spanMap.get(transactionId.meta.id).getOrElse(Nil)
+
+    val spanBuilder = tracer
+      .buildSpan(logMarker.action)
+      .withTag("transactionId", transactionId.meta.id)
+
+    val activeSpan: Span = spanList match {
+      case Nil =>
+        TracingCacheProvider.contextMap
+          .get(transactionId.meta.id)
+          .map(spanBuilder.asChildOf(_).startActive(true).span())
+          .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span())
+
+      case _ =>
+        spanList.head.get
+          .map(spanBuilder.asChildOf(_).startActive(true).span())
+          .getOrElse(spanBuilder.ignoreActiveSpan().startActive(true).span())
+
+    }
+    //add active span to list
+    TracingCacheProvider.spanMap.put(transactionId.meta.id, spanList.::(new WeakReference(activeSpan)))
+  }
+
+  /**
+   * Finish a Trace associated with given transactionId.
+   *
+   * @param transactionId
+   */
+  override def finishSpan(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Register error
+   *
+   * @param transactionId
+   */
+  override def error(transactionId: TransactionId): Unit = {
+    clear(transactionId)
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = {
+    TracingCacheProvider.spanMap.get(transactionId.meta.id) match {
+      case Some(spanList) => {
+        val map: mutable.Map[String, String] = mutable.HashMap()
+        //inject latest span context in map
+        spanList.head.get match {
+          case Some(span) => {
+            tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(map.asJava))
+            Some(map.toMap)
+          }
+          case _ => None
+        }
+      }
+      case None => None
+    }
+  }
```

Comment:
Same as above, you can use mapping and flatmapping to squash the options down, since you only need the path where the options are defined:

```scala
TracingCacheProvider.spanMap.get(transactionId.meta.id)
  .flatMap(_.headOption)
  .flatMap(_.get)
  .map { span =>
    val map = Map.empty[String, String]
    tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(map.asJava))
    map
  }
```

WDYT?